In [41]:
import pandas as pd
import numpy as np
import os
from util import HISTORICAL_FILE_JOIN, RESOURCE_DIR 
#pip install mlxtend
from mlxtend.frequent_patterns import apriori, association_rules

Reads projects from Excel file

In [25]:
#df = pd.read_excel("historical_join.xlsx", keep_default_na=False)
df = pd.read_excel(HISTORICAL_FILE_JOIN, keep_default_na=False)

len(df)
df.head()

PROJECTS  COMMITS  Memcached  Microsoft Azure Table Storage  FileMaker  \
0  Activiti        1          0                              0          0   
1  Activiti        2          0                              0          0   
2  Activiti        3          0                              0          0   
3  Activiti        4          0                              0          0   
4  Activiti        5          0                              0          0   

   Riak KV  Hive  Redis  Aerospike  Google Cloud Datastore  ...  Teradata  \
0        0     0      0          0                       0  ...         0   
1        0     0      0          0                       0  ...         0   
2        0     0      0          0                       0  ...         0   
3        0     0      0          0                       0  ...         0   
4        0     0      0          0                       0  ...         0   

   PostGIS  Etcd  MySQL  Join_Ignite-Sql  Join_Ignite-NoSql  \
0        0     0      1                0                  0   
1        0     0      1                0                  0   
2        0     0      1                0                  0   
3        0     0      1                0                  0   
4        0     0      1                0                  0   

   Join_MarkLogic-Sql  Join_MarkLogic-NoSql  Join_Virtuoso-Sql  \
0                   0                     0                  0   
1                   0                     0                  0   
2                   0                     0                  0   
3                   0                     0                  0   
4                   0                     0                  0   

   Join_Virtuoso-NoSql  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  

[5 rows x 55 columns]

Deleting PROJECTS and COMMITS columns to work with one hot values

In [45]:
df2 = df.copy()
df2

PROJECTS  COMMITS  Memcached  Microsoft Azure Table Storage  FileMaker  \
0     Activiti        1          0                              0          0   
1     Activiti        2          0                              0          0   
2     Activiti        3          0                              0          0   
3     Activiti        4          0                              0          0   
4     Activiti        5          0                              0          0   
...        ...      ...        ...                            ...        ...   
3165   maxwell        6          0                              0          0   
3166   maxwell        7          0                              0          0   
3167   maxwell        8          0                              0          0   
3168   maxwell        9          0                              0          0   
3169   maxwell       10          0                              0          0   

      Riak KV  Hive  Redis  Aerospike  Google Cloud Datastore  ...  Teradata  \
0           0     0      0          0                       0  ...         0   
1           0     0      0          0                       0  ...         0   
2           0     0      0          0                       0  ...         0   
3           0     0      0          0                       0  ...         0   
4           0     0      0          0                       0  ...         0   
...       ...   ...    ...        ...                     ...  ...       ...   
3165        0     0      0          0                       0  ...         0   
3166        0     0      1          0                       0  ...         0   
3167        0     0      1          0                       0  ...         0   
3168        0     0      1          0                       0  ...         0   
3169        0     0      0          0                       0  ...         0   

      PostGIS  Etcd  MySQL  Join_Ignite-Sql  Join_Ignite-NoSql  \
0           0     0      1                0                  0   
1           0     0      1                0                  0   
2           0     0      1                0                  0   
3           0     0      1                0                  0   
4           0     0      1                0                  0   
...       ...   ...    ...              ...                ...   
3165        0     0      1                0                  0   
3166        0     0      1                0                  0   
3167        0     0      1                0                  0   
3168        0     0      1                0                  0   
3169        0     0      1                0                  0   

      Join_MarkLogic-Sql  Join_MarkLogic-NoSql  Join_Virtuoso-Sql  \
0                      0                     0                  0   
1                      0                     0                  0   
2                      0                     0                  0   
3                      0                     0                  0   
4                      0                     0                  0   
...                  ...                   ...                ...   
3165                   0                     0                  0   
3166                   0                     0                  0   
3167                   0                     0                  0   
3168                   0                     0                  0   
3169                   0                     0                  0   

      Join_Virtuoso-NoSql  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
3165                    0  
3166                    0  
3167                    0  
3168                    0  
3169                    0  

[3170 rows x 55 columns]

In [5]:
df = 0
df

0

Agrupando as linhas para identificar quais os projetos que não possuem indício de BD
Grouping the lines to identify which projects that don't have DB clues

In [78]:
df2 = df2.groupby("PROJECTS").sum()
df2


COMMITS  Memcached  \
PROJECTS                                            
Activiti                            55          0   
AmazeFileManager                    55          0   
Android-IMSI-Catcher-Detector       55          0   
Anki-Android                        55          0   
AntennaPod                          55          0   
...                                ...        ...   
zeebe                               55          0   
zeppelin                            55          0   
zipkin                              55          0   
zookeeper                           55          0   
zotfile                             55          0   

                               Microsoft Azure Table Storage  FileMaker  \
PROJECTS                                                                  
Activiti                                                   0          0   
AmazeFileManager                                           0          0   
Android-IMSI-Catcher-Detector                              0          0   
Anki-Android                                               0          0   
AntennaPod                                                 0          0   
...                                                      ...        ...   
zeebe                                                      0          0   
zeppelin                                                   0          0   
zipkin                                                     0          0   
zookeeper                                                  0          0   
zotfile                                                    0          0   

                               Riak KV  Hive  Redis  Aerospike  \
PROJECTS                                                         
Activiti                             0     0      0          0   
AmazeFileManager                     0     0      0          0   
Android-IMSI-Catcher-Detector        0     0      0          0   
Anki-Android                         0     0      0          0   
AntennaPod                           0     0      0          0   
...                                ...   ...    ...        ...   
zeebe                                0     0      0          0   
zeppelin                             0     9      8          0   
zipkin                               0     0      0          0   
zookeeper                            0     0      0          0   
zotfile                              0     0      0          0   

                               Google Cloud Datastore  MariaDB  ...  Teradata  \
PROJECTS                                                        ...             
Activiti                                            0        0  ...         0   
AmazeFileManager                                    0        0  ...         0   
Android-IMSI-Catcher-Detector                       0        0  ...         0   
Anki-Android                                        0        0  ...         0   
AntennaPod                                          0        0  ...         0   
...                                               ...      ...  ...       ...   
zeebe                                               0        0  ...         0   
zeppelin                                           10        7  ...         0   
zipkin                                              0        7  ...         0   
zookeeper                                           0        0  ...         0   
zotfile                                             0        0  ...         0   

                               PostGIS  Etcd  MySQL  Join_Ignite-Sql  \
PROJECTS                                                               
Activiti                             0     0      6                0   
AmazeFileManager                     0     0      0                0   
Android-IMSI-Catcher-Detector        0     0      0                0   
Anki-Android                         0     0      0                0   
AntennaPod          

Removing projects that don't have DB clues 

In [ ]:
#df2.index.name = 'PROJECTS'
#df4 = df3.drop(df3[df3.iloc[: , 1:] == 0].index)
#df.rename(columns={'PROJECTS' : 'Projects'}, inplace=True)
#df2.index.name = 'PROJECTS'
#print(teste.columns)
#print(df2.columns)
#teste = teste.drop(["PROJECTS","COMMITS"], axis=1, inplace=True)
#df3 = df2[(df2 == 0).all(axis=1)]



In [80]:
#df2.index.name = 'PROJECTS'
df2 = df2.iloc[ : , 1:] #desconiderandcoluna commits 
#df3[(df3 == 0).all(axis=1) # filtra as linhas que possuem valor igual a 0 em todas as colunas
df2 = df2.drop(df2[(df2 == 0).all(axis=1)].index) # deletando as linhas que possuem valor 0 em todas as colunas
df2

Microsoft Azure Table Storage  FileMaker  \
PROJECTS                                                                  
Activiti                                                   0          0   
Android-IMSI-Catcher-Detector                              0          0   
Arduino                                                    0          0   
AsciidocFX                                                 0          0   
BIMserver                                                  0          0   
...                                                      ...        ...   
zaproxy                                                    0          0   
zeebe                                                      0          0   
zeppelin                                                   0          0   
zipkin                                                     0          0   
zookeeper                                                  0          0   

                               Riak KV  Hive  Redis  Aerospike  \
PROJECTS                                                         
Activiti                             0     0      0          0   
Android-IMSI-Catcher-Detector        0     0      0          0   
Arduino                              0     0      0          0   
AsciidocFX                           0     0      0          0   
BIMserver                            0     0      4          0   
...                                ...   ...    ...        ...   
zaproxy                              0     0      0          0   
zeebe                                0     0      0          0   
zeppelin                             0     9      8          0   
zipkin                               0     0      0          0   
zookeeper                            0     0      0          0   

                               Google Cloud Datastore  MariaDB  Neo4j  H2  \
PROJECTS                                                                    
Activiti                                            0        0      0   9   
Android-IMSI-Catcher-Detector                       0        0      0   0   
Arduino                                             0        0      0   0   
AsciidocFX                                          1        1      0   0   
BIMserver                                           0        0      0   0   
...                                               ...      ...    ...  ..   
zaproxy                                             0        0      0   0   
zeebe                                               0        0      0   0   
zeppelin                                           10        7      4   7   
zipkin                                              0        7      0   3   
zookeeper                                           0        0      0   0   

                               ...  Teradata  PostGIS  Etcd  MySQL  \
PROJECTS                       ...                                   
Activiti                       ...         0        0     0      6   
Android-IMSI-Catcher-Detector  ...         0        0     0      0   
Arduino                        ...         0        0     0      0   
AsciidocFX                     ...         0        0     0      1   
BIMserver                      ...         0        0     0     10   
...                            ...       ...      ...   ...    ...   
zaproxy                        ...         0        0     0      6   
zeebe                          ...         0        0     0      0   
zeppelin                       ...         0        0     0      8   
zipkin                         ...         0        0     0      9   
zookeeper                      ...         0        0     0      2   

                               Join_Ignite-Sql  Join_Ignite-NoSql  \
PROJECTS                                                            
Activiti                                     0                  0   
Android-IMSI-Catcher-Detector                0                  0   
Arduino

Changing all values ​​greater than 1 to 1 because the mlextend library requires a hotvalues ​​dataset

In [81]:
df2 = df2.mask(df2 > 1, 1)
df2

Microsoft Azure Table Storage  FileMaker  \
PROJECTS                                                                  
Activiti                                                   0          0   
Android-IMSI-Catcher-Detector                              0          0   
Arduino                                                    0          0   
AsciidocFX                                                 0          0   
BIMserver                                                  0          0   
...                                                      ...        ...   
zaproxy                                                    0          0   
zeebe                                                      0          0   
zeppelin                                                   0          0   
zipkin                                                     0          0   
zookeeper                                                  0          0   

                               Riak KV  Hive  Redis  Aerospike  \
PROJECTS                                                         
Activiti                             0     0      0          0   
Android-IMSI-Catcher-Detector        0     0      0          0   
Arduino                              0     0      0          0   
AsciidocFX                           0     0      0          0   
BIMserver                            0     0      1          0   
...                                ...   ...    ...        ...   
zaproxy                              0     0      0          0   
zeebe                                0     0      0          0   
zeppelin                             0     1      1          0   
zipkin                               0     0      0          0   
zookeeper                            0     0      0          0   

                               Google Cloud Datastore  MariaDB  Neo4j  H2  \
PROJECTS                                                                    
Activiti                                            0        0      0   1   
Android-IMSI-Catcher-Detector                       0        0      0   0   
Arduino                                             0        0      0   0   
AsciidocFX                                          1        1      0   0   
BIMserver                                           0        0      0   0   
...                                               ...      ...    ...  ..   
zaproxy                                             0        0      0   0   
zeebe                                               0        0      0   0   
zeppelin                                            1        1      1   1   
zipkin                                              0        1      0   1   
zookeeper                                           0        0      0   0   

                               ...  Teradata  PostGIS  Etcd  MySQL  \
PROJECTS                       ...                                   
Activiti                       ...         0        0     0      1   
Android-IMSI-Catcher-Detector  ...         0        0     0      0   
Arduino                        ...         0        0     0      0   
AsciidocFX                     ...         0        0     0      1   
BIMserver                      ...         0        0     0      1   
...                            ...       ...      ...   ...    ...   
zaproxy                        ...         0        0     0      1   
zeebe                          ...         0        0     0      0   
zeppelin                       ...         0        0     0      1   
zipkin                         ...         0        0     0      1   
zookeeper                      ...         0        0     0      1   

                               Join_Ignite-Sql  Join_Ignite-NoSql  \
PROJECTS                                                            
Activiti                                     0                  0   
Android-IMSI-Catcher-Detector                0                  0   
Arduino

In [82]:
df2 = df2.reset_index()
df2

PROJECTS  Microsoft Azure Table Storage  FileMaker  \
0                         Activiti                              0          0   
1    Android-IMSI-Catcher-Detector                              0          0   
2                          Arduino                              0          0   
3                       AsciidocFX                              0          0   
4                        BIMserver                              0          0   
..                             ...                            ...        ...   
192                        zaproxy                              0          0   
193                          zeebe                              0          0   
194                       zeppelin                              0          0   
195                         zipkin                              0          0   
196                      zookeeper                              0          0   

     Riak KV  Hive  Redis  Aerospike  Google Cloud Datastore  MariaDB  Neo4j  \
0          0     0      0          0                       0        0      0   
1          0     0      0          0                       0        0      0   
2          0     0      0          0                       0        0      0   
3          0     0      0          0                       1        1      0   
4          0     0      1          0                       0        0      0   
..       ...   ...    ...        ...                     ...      ...    ...   
192        0     0      0          0                       0        0      0   
193        0     0      0          0                       0        0      0   
194        0     1      1          0                       1        1      1   
195        0     0      0          0                       0        1      0   
196        0     0      0          0                       0        0      0   

     ...  Teradata  PostGIS  Etcd  MySQL  Join_Ignite-Sql  Join_Ignite-NoSql  \
0    ...         0        0     0      1                0                  0   
1    ...         0        0     0      0                0                  0   
2    ...         0        0     0      0                0                  0   
3    ...         0        0     0      1                0                  0   
4    ...         0        0     0      1                0                  0   
..   ...       ...      ...   ...    ...              ...                ...   
192  ...         0        0     0      1                0                  0   
193  ...         0        0     0      0                0                  0   
194  ...         0        0     0      1                0                  1   
195  ...         0        0     0      1                0                  0   
196  ...         0        0     0      1                0                  0   

     Join_MarkLogic-Sql  Join_MarkLogic-NoSql  Join_Virtuoso-Sql  \
0                     0                     0                  0   
1                     0                     0                  0   
2                     0                     0                  0   
3                     0                     0                  0   
4                     0                     0                  0   
..                  ...                   ...                ...   
192                   0                     0                  0   
193                   0                     0                  0   
194                   0                     0                  0   
195                   0                     0                  0   
196                   0                     0                  0   

     Join_Virtuoso-NoSql  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
192                    0  
193                    0  
194                    0  
195                    0  
196                    0  

[

In [83]:
df2.drop(["PROJECTS"], axis=1, inplace=True)
df2


Microsoft Azure Table Storage  FileMaker  Riak KV  Hive  Redis  \
0                                0          0        0     0      0   
1                                0          0        0     0      0   
2                                0          0        0     0      0   
3                                0          0        0     0      0   
4                                0          0        0     0      1   
..                             ...        ...      ...   ...    ...   
192                              0          0        0     0      0   
193                              0          0        0     0      0   
194                              0          0        0     1      1   
195                              0          0        0     0      0   
196                              0          0        0     0      0   

     Aerospike  Google Cloud Datastore  MariaDB  Neo4j  H2  ...  Teradata  \
0            0                       0        0      0   1  ...         0   
1            0                       0        0      0   0  ...         0   
2            0                       0        0      0   0  ...         0   
3            0                       1        1      0   0  ...         0   
4            0                       0        0      0   0  ...         0   
..         ...                     ...      ...    ...  ..  ...       ...   
192          0                       0        0      0   0  ...         0   
193          0                       0        0      0   0  ...         0   
194          0                       1        1      1   1  ...         0   
195          0                       0        1      0   1  ...         0   
196          0                       0        0      0   0  ...         0   

     PostGIS  Etcd  MySQL  Join_Ignite-Sql  Join_Ignite-NoSql  \
0          0     0      1                0                  0   
1          0     0      0                0                  0   
2          0     0      0                0                  0   
3          0     0      1                0                  0   
4          0     0      1                0                  0   
..       ...   ...    ...              ...                ...   
192        0     0      1                0                  0   
193        0     0      0                0                  0   
194        0     0      1                0                  1   
195        0     0      1                0                  0   
196        0     0      1                0                  0   

     Join_MarkLogic-Sql  Join_MarkLogic-NoSql  Join_Virtuoso-Sql  \
0                     0                     0                  0   
1                     0                     0                  0   
2                     0                     0                  0   
3                     0                     0                  0   
4                     0                     0                  0   
..                  ...                   ...                ...   
192                   0                     0                  0   
193                   0                     0                  0   
194                   0                     0                  0   
195                   0                     0                  0   
196                   0                     0                  0   

     Join_Virtuoso-NoSql  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
192                    0  
193                    0  
194                    0  
195                    0  
196                    0  

[197 rows x 52 columns]

In [15]:
df2 = df2.astype(np.uint8)

In [84]:
df2

Microsoft Azure Table Storage  FileMaker  Riak KV  Hive  Redis  \
0                                0          0        0     0      0   
1                                0          0        0     0      0   
2                                0          0        0     0      0   
3                                0          0        0     0      0   
4                                0          0        0     0      1   
..                             ...        ...      ...   ...    ...   
192                              0          0        0     0      0   
193                              0          0        0     0      0   
194                              0          0        0     1      1   
195                              0          0        0     0      0   
196                              0          0        0     0      0   

     Aerospike  Google Cloud Datastore  MariaDB  Neo4j  H2  ...  Teradata  \
0            0                       0        0      0   1  ...         0   
1            0                       0        0      0   0  ...         0   
2            0                       0        0      0   0  ...         0   
3            0                       1        1      0   0  ...         0   
4            0                       0        0      0   0  ...         0   
..         ...                     ...      ...    ...  ..  ...       ...   
192          0                       0        0      0   0  ...         0   
193          0                       0        0      0   0  ...         0   
194          0                       1        1      1   1  ...         0   
195          0                       0        1      0   1  ...         0   
196          0                       0        0      0   0  ...         0   

     PostGIS  Etcd  MySQL  Join_Ignite-Sql  Join_Ignite-NoSql  \
0          0     0      1                0                  0   
1          0     0      0                0                  0   
2          0     0      0                0                  0   
3          0     0      1                0                  0   
4          0     0      1                0                  0   
..       ...   ...    ...              ...                ...   
192        0     0      1                0                  0   
193        0     0      0                0                  0   
194        0     0      1                0                  1   
195        0     0      1                0                  0   
196        0     0      1                0                  0   

     Join_MarkLogic-Sql  Join_MarkLogic-NoSql  Join_Virtuoso-Sql  \
0                     0                     0                  0   
1                     0                     0                  0   
2                     0                     0                  0   
3                     0                     0                  0   
4                     0                     0                  0   
..                  ...                   ...                ...   
192                   0                     0                  0   
193                   0                     0                  0   
194                   0                     0                  0   
195                   0                     0                  0   
196                   0                     0                  0   

     Join_Virtuoso-NoSql  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
192                    0  
193                    0  
194                    0  
195                    0  
196                    0  

[197 rows x 52 columns]

Executing an a priori function to extract common itemsets for association rule acquisition
Support = 0,01 corresponds to 19.7 tuples from our dataset (197 records).
This means that only item sets that occurred more than 19.7 times in our data will be considered.

In [85]:
frq_items = apriori(df2, min_support = 0.021, use_colnames = True)

Collecting the inferred rules in a dataframe whose lift value is above 1.

In [86]:
rules = association_rules(frq_items, metric ="lift", min_threshold = 1)

In [87]:
#rules = association_rules(frq_items, metric ="lift", min_threshold = 1)

# Sorting the rules by the lift measure 
rules = rules.sort_values('lift', ascending =False)
print(rules.head())

                                              antecedents  \
192821                         (Amazon Redshift, IBM DB2)   
288692        (MariaDB, Amazon Redshift, IBM DB2, Oracle)   
615572  (Redis, Cassandra, MongoDB, SAP Adaptive Serve...   
350483  (PostgreSQL, MySQL, SAP Adaptive Server, Cassa...   
615569  (Redis, Cassandra, Elasticsearch, SAP Adaptive...   

                                           consequents  antecedent support  \
192821   (MySQL, Snowflake, PostgreSQL, MS SQL Server)            0.025381   
288692          (Snowflake, PostgreSQL, MS SQL Server)            0.025381   
615572  (MariaDB, MySQL, Oracle, Elasticsearch, HBase)            0.025381   
350483                 (MariaDB, HBase, Oracle, Redis)            0.025381   
615569        (MariaDB, MySQL, Oracle, MongoDB, HBase)            0.025381   

        consequent support   support  confidence  lift  leverage  conviction  
192821            0.025381  0.025381         1.0  39.4  0.024737         inf  
288692

Showing all generated association rules

In [88]:
rules

antecedents  \
192821                         (Amazon Redshift, IBM DB2)   
288692        (MariaDB, Amazon Redshift, IBM DB2, Oracle)   
615572  (Redis, Cassandra, MongoDB, SAP Adaptive Serve...   
350483  (PostgreSQL, MySQL, SAP Adaptive Server, Cassa...   
615569  (Redis, Cassandra, Elasticsearch, SAP Adaptive...   
...                                                   ...   
381                                           (Hazelcast)   
4842                                            (Ehcache)   
4839                                 (SQLite, PostgreSQL)   
9181                                              (Redis)   
9168                          (SQLite, MySQL, PostgreSQL)   

                                           consequents  antecedent support  \
192821   (MySQL, Snowflake, PostgreSQL, MS SQL Server)            0.025381   
288692          (Snowflake, PostgreSQL, MS SQL Server)            0.025381   
615572  (MariaDB, MySQL, Oracle, Elasticsearch, HBase)            0.025381   
350483                 (MariaDB, HBase, Oracle, Redis)            0.025381   
615569        (MariaDB, MySQL, Oracle, MongoDB, HBase)            0.025381   
...                                                ...                 ...   
381                                            (MySQL)            0.106599   
4842                              (SQLite, PostgreSQL)            0.289340   
4839                                         (Ehcache)            0.116751   
9181                       (SQLite, MySQL, PostgreSQL)            0.309645   
9168                                           (Redis)            0.111675   

        consequent support   support  confidence       lift  leverage  \
192821            0.025381  0.025381    1.000000  39.400000  0.024737   
288692            0.025381  0.025381    1.000000  39.400000  0.024737   
615572            0.025381  0.025381    1.000000  39.400000  0.024737   
350483            0.025381  0.025381    1.000000  39.400000  0.024737   
615569            0.025381  0.025381    1.000000  39.400000  0.024737   
...                    ...       ...         ...        ...       ...   
381               0.543147  0.060914    0.571429   1.052069  0.003015   
4842              0.116751  0.035533    0.122807   1.051869  0.001752   
4839              0.289340  0.035533    0.304348   1.051869  0.001752   
9181              0.111675  0.035533    0.114754   1.027571  0.000953   
9168              0.309645  0.035533    0.318182   1.027571  0.000953   

        conviction  
192821         inf  
288692         inf  
615572         inf  
350483         inf  
615569         inf  
...            ...  
381       1.065990  
4842      1.006904  
4839      1.021574  
9181      1.003478  
9168      1.012521  

[712094 rows x 9 columns]

Counting the size of each rule's itemsets 

In [89]:
rules['antecedent_len'] = rules['antecedents'].apply(lambda x: len(x))
rules['consequente_len'] = rules['consequents'].apply(lambda x: len(x))

Filtering only the rules of size 2, that is, 1 antecedent and 1 consequent. 

In [90]:
rules_filter = rules[(rules['antecedent_len'] == 1) & (rules['consequente_len'] == 1)]

Showing all filtered rules.

In [91]:
rules_filter

antecedents        consequents  antecedent support  \
407        (Snowflake)  (Amazon Redshift)            0.030457   
406  (Amazon Redshift)        (Snowflake)            0.030457   
237          (SapHana)         (Firebird)            0.030457   
236         (Firebird)          (SapHana)            0.040609   
238         (Informix)          (SapHana)            0.045685   
..                 ...                ...                 ...   
356           (SQLite)          (MongoDB)            0.182741   
388           (SQLite)          (Ehcache)            0.182741   
389          (Ehcache)           (SQLite)            0.289340   
380            (MySQL)        (Hazelcast)            0.543147   
381        (Hazelcast)            (MySQL)            0.106599   

     consequent support   support  confidence       lift  leverage  \
407            0.030457  0.025381    0.833333  27.361111  0.024453   
406            0.030457  0.025381    0.833333  27.361111  0.024453   
237            0.040609  0.025381    0.833333  20.520833  0.024144   
236            0.030457  0.025381    0.625000  20.520833  0.024144   
238            0.030457  0.025381    0.555556  18.240741  0.023989   
..                  ...       ...         ...        ...       ...   
356            0.182741  0.035533    0.194444   1.064043  0.002139   
388            0.289340  0.055838    0.305556   1.056043  0.002963   
389            0.182741  0.055838    0.192982   1.056043  0.002963   
380            0.106599  0.060914    0.112150   1.052069  0.003015   
381            0.543147  0.060914    0.571429   1.052069  0.003015   

     conviction  antecedent_len  consequente_len  
407    5.817259               1                1  
406    5.817259               1                1  
237    5.756345               1                1  
236    2.585448               1                1  
238    2.181472               1                1  
..          ...             ...              ...  
356    1.014528               1                1  
388    1.023350               1                1  
389    1.012690               1                1  
380    1.006252               1                1  
381    1.065990               1                1  

[476 rows x 11 columns]

Generating the dataframe with the filtered rules.

In [92]:
HISTORICAL_RULES_FILE = RESOURCE_DIR + os.sep + 'seq_patterns/testes/historical_rules_197_021_2.xlsx'
rules_filter = rules_filter.to_excel(HISTORICAL_RULES_FILE, index=False)